pkg/metrics/cluster/node.go

```diff
@@ -176,59 +175,66 @@ func (n *node) run() {
 func (n *node) performClusterReshard(ctx context.Context, joining bool) error {
 	if n.ring == nil || n.lc == nil {
 		level.Info(n.log).Log("msg", "node disabled, not resharding")
-		return nil
-	}
-
-	if n.cfg.ReshardTimeout > 0 {
-		var cancel context.CancelFunc
-		ctx, cancel = context.WithTimeout(ctx, n.cfg.ReshardTimeout)
-		defer cancel()
 	}
 
 	var (
 		rs  ring.ReplicationSet
 		err error
-
-		firstError error
 	)
 
+	// This cancel has to exist outside since it is used in the goroutine
+	// originally it was put in a if !joining block which would logically
+	// ensure it would be called, but the linter does not process those paths
+	// so instead its sprinkled around before the returns.
+	var cancel context.CancelFunc
+	if n.cfg.RefreshTimeout > 0 {
+		ctx, cancel = context.WithTimeout(ctx, n.cfg.RefreshTimeout)
+	}
+
 	backoff := cortex_util.NewBackoff(ctx, backoffConfig)
 	for backoff.Ongoing() {
+		if ctx.Err() != nil {
+			cancel()
+			return ctx.Err()
+		}
 		rs, err = n.ring.GetAllHealthy(ring.Read)
 		if err == nil {
 			break
 		}
 		backoff.Wait()
 	}
-	if err := backoff.Err(); err != nil && firstError == nil {
-		firstError = err
-	}
 
 	if len(rs.Instances) > 0 {
 		level.Info(n.log).Log("msg", "informing remote nodes to reshard")
 	}
 
+	// These are not in the go routine below due to potential race condition with n.lc.addr
 	_, err = rs.Do(ctx, 500*time.Millisecond, func(c context.Context, id *ring.InstanceDesc) (interface{}, error) {
 		// Skip over ourselves.
 		if id.Addr == n.lc.Addr {
 			return nil, nil
 		}
 
-		ctx = user.InjectOrgID(ctx, "fake")
-		return nil, n.notifyReshard(ctx, id)
+		notifyCtx := user.InjectOrgID(c, "fake")
+		return nil, n.notifyReshard(notifyCtx, id)
 	})
-	if err != nil && firstError == nil {
-		firstError = err
+
+	if err != nil {
+		level.Error(n.log).Log("msg", "notifying other nodes failed", "err", err)
 	}
 
+	// This is ran in the background to help prevent unnecessary mutex locking.
 	if joining {
-		level.Info(n.log).Log("msg", "running local reshard")
-		if _, err := n.srv.Reshard(ctx, &pb.ReshardRequest{}); err != nil {
-			level.Warn(n.log).Log("msg", "dynamic local reshard did not succeed", "err", err)
-		}
+		go func() {
+			defer cancel()
+			level.Info(n.log).Log("msg", "running local reshard")
+			if _, reshardErr := n.srv.Reshard(ctx, &pb.ReshardRequest{}); reshardErr != nil {
+				level.Warn(n.log).Log("msg", "dynamic local reshard did not succeed", "err", err)
+			}
+		}()
 	}
-
-	return firstError
+	cancel()
+	return err
```

Comment:
This chunk here is still subject to timeout issues as we discussed offline. I strongly think we need to have the implementation of `Reshard` to kick off a Goroutine, and remove spawning a Goroutine here. 

As it is now, the calls to `.notifyReshard` will block until all other nodes finish refreshing, which may also include waiting for one of the tick refreshes to finish. It would lead to a similar failure issue with the new timeout we added.

Reply:
Per our conversation moved to the reshard function. Tested to see if any negative effects would occur and none were found.